### DISCUSSION TITLE
Best way to add aarch64 support?

### DISCUSSION THREAD

**Discussion by jvoisin** [fwupd/fwupd#7310] [Q&A] [answered]
I'm currently toying with `fwupdmgr security --force` on an Apple M2 running AsahiLinux, and I'm getting `✘ Supported CPU:                 Unknown`, which is expected since fwupd [is using `cpuid`](https://github.com/fwupd/fwupd/blob/main/libfwupdplugin/fu-common.c#L38) to get data about the cpu's vendor, which is an x86 instruction.

What would be the best way to add support for aarch64? I think the less worse way would be to abstract things beneath `get_vendor`/`get_model`/`get_extended_features`/`get_manufacturer`/… functions with `#ifdef` in them, using [msr](https://www.kernel.org/doc/html/v5.6/arm64/cpu-feature-registers.html) for aarch64.

If this sounds like a good idea, I can start trying to implement it.

**Comment by superm1** [comment]
Detecting the CPU I think is a (relatively) trivial problem.
To avoid putting cart before the horse it's probably a good idea to talk about what kind of checks are going to _make sense_ in this silicon.

You can see what's in place today at the spec here:
https://fwupd.github.io/libfwupdplugin/hsi.html

**Reply by jvoisin** [reply]
I'd say at least:
- kernel Lockdown
- BTI
- PAC
- MTE
- DMA/IOMMU
 
As well as [other Apple stuff](https://support.apple.com/guide/security/operating-system-integrity-sec8b776536b/web)

**Reply by hughsie** [reply, marked as answer]
Sure, I think phase 1 is probably to add HSI checks for those last 4 bullet points. Like @superm1 the CPUID check is trivial -- it's the others that are slightly harder to implement. I've got no political problem enabling mac hardware for HSI tests, as I've been really impressed with what the AsahiLinux team have been able to pull off and if it's a end-user useful set of tests I think it makes sense. Actually writing a fwupd plugin that checks something like BTU/PAC/MTE is probably only a few minutes work, assuming the data is available to usespace. If you need to add mac-specific attributes (and I think you do) then we just need to write the spec element in https://github.com/fwupd/fwupd/tree/main/docs/hsi-tests.d -- yell if you need any help and we can give advice and sample code.

**Reply by jvoisin** [reply]
I'd love some hints on the best way to make the code clean:

- is a bunch of `#ifdef` in get_vendor/get_model/get_extended_features/get_manufacturer/… the way to go?
- there are currently some information missing from the cpuinfo compared to x84_64, shall I use other sources than the "cpuid instruction equivalent" to try to fill the processor_id/family_id/model_id/stepping_id/… field?

**Reply by superm1** [reply]
IMO I think we should move some code to a file that is only compiled for a given architecture (like a "backend") and leave a "high" level interface for the plugin to use.

It's early for it; but I want to put the idea in your mind too that we should have CI for aarch64.  Cross compile + QEMU, QEMU build and test; remote host to run it, however it can be done.  Otherwise we're going to end up with code that doesn't have proper test coverage.

**Comment by marcan** [comment]
Just to make sure we're all on the same page here:

- The Apple specific stuff is largely macOS-specific and does not apply to Linux (no kernel integrity protection, no fast permission restrictions, SCIP is questionably implemented from what I've seen in our context, no PPL, and there is no chance of this ever being implemented for Linux since it would be highly intrusive, based on proprietary features, and upstream would reject it). Just PAC/BTI (which are generic features) and of course the IOMMU stuff.
- We do intend to support "proper" Linux secureboot on these machines (and auditing for that makes some sense) but since it doesn't exist yet nor is an implementation decided upon, there is nothing to test for right now.
- Just for clarity, there is probably no place for fwupd to update any machine firmware on these machines. The systems are special enough that, once we figure out how to do this at all, it will likely be implemented as a very bespoke stand-alone tool that integrates into Apple's firmware update flow. Of course you could still use fwupd to update accessory/external device firmware.

**Reply by hughsie** [reply]
> it will likely be implemented as a very bespoke stand-alone tool

Sounds like a fwupd plugin to me :) Then you get full integration into the GUI tools for free.

**Reply by marcan** [reply]
The firmware updates involve downloading files from Apple's CDN, setting up files in APFS partitions, and rebooting into special firmware update ramdisks. There's actually a good chance we end up triggering this entirely from recoveryOS (which is a stripped down macOS) and not from Linux. After that you need to update the Asahi Linux stub partition, which **requires** a trip through recoveryOS and the user entering their machine owner credentials. Plus we need to carefully gate it on kernel version, because doing the latter step with a kernel that doesn't support that OS-firmware version will cause lots of things to break.

It's a different world in Apple land.

**Reply by hughsie** [reply]
> It's a different world in Apple land.

Wowsa, indeed.
